### PR TITLE
feat: we get hydra-admin services

### LINKED ISSUE #11: Call API to get entrypoints core documentation
See [doc](https://www.w3.org/ns/hydra/core)

### PR DESCRIPTION
**Pull request by Simperfit**
Fix #11 #10 #12 

This has to be merger after #23 in master

### CODE REVIEW

File: config/karma.conf.js [hunk 1]
```diff
@@ -21,6 +21,8 @@ var configuration = {
     { pattern: 'dist/vendor/systemjs/dist/system-polyfills.js', included: true, watched: false },
     { pattern: 'dist/vendor/systemjs/dist/system.src.js', included: true, watched: false },
     { pattern: 'dist/vendor/zone.js/dist/async-test.js', included: true, watched: false },
+    { pattern: 'dist/vendor/zone.js/dist/fake-async-test.js', included: true, watched: false },
+
```
Comment (Gregcop1) [inline config/karma.conf.js]: useless blank line

File: src/app/hm-admin/helper/entrypointHelper.spec.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import {it, describe, expect, beforeEach} from '@angular/core/testing';
+import {COLLECTION_METHOD, entrypointHelper} from './entrypointHelper';
+import {EntryPoint} from '../services/models/EntryPoint';
+
```
Comment (Gregcop1) [inline src/app/hm-admin/helper/entrypointHelper.spec.ts:4]: several useless blank lines

File: src/app/hm-admin/helper/entrypointHelper.spec.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      it('should get operations', () => {
+        let operations = [
+          {'hydra:method': 'GET'},
+          {'hydra:method': 'POST'}
+        ],
+          results = entrypointHelper.getOperations(operations);
+
+        expect(results.length).toBe(2);
+        expect(results[0]).toBe('GET');
```
Comment (Gregcop1) [inline src/app/hm-admin/helper/entrypointHelper.spec.ts]: maybe using `toContain`instead

File: src/app/hm-admin/helper/entrypointHelper.spec.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      });
+
+      it('should detect get collection operation', () => {
+        let operations = [
+            {'hydra:method': 'GET', 'returns': 'hydra:PagedCollection'}
+          ],
+          results = entrypointHelper.getOperations(operations);
+
+        expect(results.length).toBe(1);
+        expect(results[0]).toBe(COLLECTION_METHOD);
```
Comment (Gregcop1) [inline src/app/hm-admin/helper/entrypointHelper.spec.ts]: `toContain`

File: src/app/hm-admin/helper/entrypointHelper.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+   * Get Operations
+   *
+   * @param {Array<any>} supportedOperations
+   *
+   * @returns {Array<string>}
+   * @private
+   */
+    getOperations(supportedOperations: Array<any>): Array<string> {
+    return _.map(supportedOperations, ((operation: any) => {
+      return (operation['hydra:method'] && operation.returns === 'hydra:PagedCollection') ? COLLECTION_METHOD : operation['hydra:method'];
```
Comment (Gregcop1) [inline src/app/hm-admin/helper/entrypointHelper.ts]: should be definition

File: src/app/hm-admin/services/models/PagedCollection.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  public context;
+  public id;
+  public type;
+  public firstPage;
+  public itemsPerPage;
+  public lastPage;
+  public members;
+  public totalItems;
+
+  constructor(properties: any = {}) {
```
Comment (Gregcop1) [inline src/app/hm-admin/services/models/PagedCollection.ts]: should use definitions

**Comment by Gregcop1** [comment]
change the label. I've too much code in this to validate it so let's @gorghoa do the job

File: src/app/hm-admin/helper/schemaHelper.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+   * Clean the list of model from polluted datas
+   *
+   * @param supportedClasses
+   *
+   * @returns {Array<any>}
+   * @private
+   */
+    cleanModelsList(supportedClasses: Array<any>): Array<any> {
+    if (undefined !== supportedClasses) {
+      return supportedClasses.slice(0, -3);
```
Comment (gorghoa) [inline src/app/hm-admin/helper/schemaHelper.ts]: Are we hundred percent sure of this slice length?

Comment (Gregcop1) [inline src/app/hm-admin/helper/schemaHelper.ts]: Hmm, you're right, but it seem so.
In API Platform there's Entrypoints, ConstraintViolation and ConstraintViolationList ([vocab](https://api-platform-demo-blog-api.herokuapp.com/vocab))
But in pure hydra, it's Event, vocab:EntryPoint, ConstraintViolationList ([vocab](http://www.markus-lanthaler.com/hydra/event-api/vocab))

@Simperfit maybe you should find a better filter

Comment (Simperfit) [inline src/app/hm-admin/helper/schemaHelper.ts]: What should I do on this ?

Comment (Gregcop1) [inline src/app/hm-admin/helper/schemaHelper.ts]: Analyse requirement of hydra, of API Platform, reflect, innovate, be creative and find a solution. And more important. Trust yourself

Comment (Gregcop1) [inline src/app/hm-admin/helper/schemaHelper.ts]: Maybe we should only keep @id which start by 'http://schema.org/'

**Comment by Simperfit** [comment]
ping @coopTilleuls/hm-admin this need to be reviewed

File: src/app/hm-admin/helper/entrypointHelper.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+export let entrypointHelper = {
+  /**
+   * @param {any} datas
+   *
+   * @returns {any} datas
+   * @private
+   */
+    filterEntryPoints(datas: any): any {
+    return _.omit(datas, ['@context', '@id', '@type']);
+  },
```
Comment (gorghoa) [inline src/app/hm-admin/helper/entrypointHelper.ts]: curious indent

File: src/app/hm-admin/services/api/api.service.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+   * @param {string} path
+   *
+   * @returns {Observable<PagedCollection|null>}
+   */
+  getCollectionByUrl(path: string, definition: any) {
+    let query = this.http.get(path)
+      .map(data => data.json())
+      .map(collection => new PagedCollection(collection, definition));
+
+    query.subscribe(
```
Comment (gorghoa) [inline src/app/hm-admin/services/api/api.service.ts:36]: Doesn’t `catch` is a better option than  subscribing? @Gregcop1, any thoughts?

File: src/app/hm-admin/services/api/api.service.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  /**
+   * Get collection from API
+   *
+   * @param {string} modelName
+   * @returns {Observable<PagedCollection|null>}
+   */
+  getCollection(modelName: string = '') {
+    let cleanName: string = stringHelper.toCamelCase(modelName),
+      entrypoint: EntryPoint = entrypointHelper.getEntryPointByModel(cleanName, this.entrypoints);
+    if (undefined === entrypoint || '' === modelName) {
```
Comment (gorghoa) [inline src/app/hm-admin/services/api/api.service.ts:55]: warning, I bet you should use : `'undefined' === typeof entrypoint`.

Moreover, is “undefined” what you really want to check? Isn’t a simple `!entrypoint` a safer choice (still weak, I admit) ? Considering what you really want is `entrypoint.url`.

Comment (Gregcop1) [inline src/app/hm-admin/services/api/api.service.ts:55]: why ?

Comment (gorghoa) [inline src/app/hm-admin/services/api/api.service.ts:55]: Good point, given the circumstances, comparing with `undefined` is unlikely to throws… 
Unless we’d check `entrypoint.url`, `undefined === entrypoint` is ok…

My point was also that entrypoint could be of any sort of type and would pass the test successfully if set, but will mess up the following part of the function, which assumes `entrypoint.url` is also set.

File: src/app/hm-admin/services/models/PagedCollection.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  public context;
+  public id;
+  public type;
+  public firstPage;
+  public itemsPerPage;
+  public lastPage;
+  public members;
+  public totalItems;
+
+  constructor(properties: any = {}, definition: any = {}) {
```
Comment (gorghoa) [inline src/app/hm-admin/services/models/PagedCollection.ts:12]: Since we are assuming that properties has some well known properties of its own (`@context`, `@id`, etc.). We should have a typescript interface for it.

File: src/app/hm-admin/services/schema/schema.service.ts [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  loadRoot() {
+    let request: string = this.config.api.baseUrl;
+    let entrypoints = this.http.get(request)
+      .map(data => data.json())
+      .map(entrypointHelper.filterEntryPoints);
+
+    entrypoints.subscribe(
+      (entrypoints => {
+        _.forEach(_.keys(entrypoints), (key) => {
```
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts]: Is there a reason for using lodash instead of `Object.keys`, or even directly `for const key in entrypoints`?

File: src/app/hm-admin/services/schema/schema.service.ts [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+            entrypoint = new EntryPoint(cleanModel);
+            this.availableEntrypoints.push(entrypoint);
+          }
+
+          entrypoint.url = entrypoints[key];
+        });
+        this.entrypoints.next(this.availableEntrypoints);
+      }),
+      () => {
+        throw new Error(`API's url is unreachable: ${request}.`);
```
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts]: Aren’t we losing the original error by simply ignoring it?

File: src/app/hm-admin/services/schema/schema.service.ts [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+      }
+    );
+  }
+
+  /**
+   * load documentation about the API
+   */
+  loadDocumentation(definition: any) {
+    let request = [this.config.api.baseUrl, 'vocab'].join('/'),
```
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts:81]: Use the definition url from config

Comment (Simperfit) [inline src/app/hm-admin/services/schema/schema.service.ts:81]: This is not the same thing. If we want to make it configurable, it's the whole line that we should add to the configuration.

File: src/app/hm-admin/services/schema/schema.service.ts (hunk 13 above)
Comment (Simperfit) [inline src/app/hm-admin/services/schema/schema.service.ts]: We are not, if the API is really unreachable, you will get this error. Trust me with that :p

File: src/app/hm-admin/services/schema/schema.service.ts [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  /**
+   * Load collection operations
+   *
+   * @param {Observable<any>} api
+   * @param {string} request
+   * @private
+   */
+  loadCollectionOperations(api: Observable<any>, definition: any) {
```
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts:110]: We should define an interface for the objects carried by the api’s parameter observable since it seems we know them well (we are using the `@id` property of such object on the following lines.). 

Same goes for the definition

File: src/app/hm-admin/services/schema/schema.service.ts [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+  /**
+   * Load single operations
+   *
+   * @param {Observable<any>} api
+   * @param {any} definition
+   * @param {string} request
+   * @private
+   */
+  loadSingleOperations(api: Observable<any>, definition: any) {
```
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts:141]: as stated before, we should use interfaces.

File: src/app/hm-admin/services/schema/schema.service.ts (hunk 13 above)
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts]: I do trust you, but:

> Aren’t we losing the original error?

File: src/app/hm-admin/services/schema/schema.service.ts (hunk 14 above)
Comment (gorghoa) [inline src/app/hm-admin/services/schema/schema.service.ts:81]: Are you confident that _all_ APIs will have this `vocab` thing? Seems risky to me.

Comment (Simperfit) [inline src/app/hm-admin/services/schema/schema.service.ts:81]: All Hydra api will.
Anyway you are right, we have to pass this in the configService

File: src/app/hm-admin/services/api/api.service.ts (hunk 10 above)
Comment (Gregcop1) [inline src/app/hm-admin/services/api/api.service.ts:55]: no it can't. 
`entrypoint: Entrypoint` => So If it's neither undefined or Entrypoint, it throw an Error

File: src/app/hm-admin/services/models/PagedCollection.ts (hunk 11 above)
Comment (Gregcop1) [inline src/app/hm-admin/services/models/PagedCollection.ts:12]: Yes. Btw, this PagedCollection should be in the List PR, not here, but we will complete it later

File: src/app/hm-admin/services/schema/schema.service.ts (hunk 13 above)
Comment (Gregcop1) [inline src/app/hm-admin/services/schema/schema.service.ts]: If you find this error is not enough complete, we have to complete it. 
I'm not sure but I think if you encounter a 404, you will have both errors

Comment (Simperfit) [inline src/app/hm-admin/services/schema/schema.service.ts]: You will get both in the console (the HTTP error with the headers, content, error code...) and this one.

File: src/app/hm-admin/services/api/api.service.ts (hunk 9 above)
Comment (Gregcop1) [inline src/app/hm-admin/services/api/api.service.ts:36]: the second parameter of subscribe is a catch.
It seems that the [catch operator](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/catch.md) is only used to catch errors until a good result come.

File: src/app/hm-admin/services/schema/schema.service.ts (hunk 12 above)
Comment (Gregcop1) [inline src/app/hm-admin/services/schema/schema.service.ts]: Good point

File: src/app/hm-admin/services/api/api.service.ts (hunk 10 above)
Comment (gorghoa) [inline src/app/hm-admin/services/api/api.service.ts:55]: It will throw at compile time, not runtime. I may have not enough xp to be comfortable with the ins and outs, so lets say I trust you on this @Gregcop1 ;)

Comment (Gregcop1) [inline src/app/hm-admin/services/api/api.service.ts:55]: Oh ok I got your point. But `'undefined' === typeof entrypoint` doesn't test that neither. If you want to test it let's play with `instanceof Entrypoint`. But do we have to test it on each function, even if they're private functions and we control input and output on each function ?

Comment (gorghoa) [inline src/app/hm-admin/services/api/api.service.ts:55]: Nope, but if we just want to check if entrypoint exists, why just not use `!entrypoint` instead of the longer and the meaningless undefined check?

File: src/app/hm-admin/services/schema/schema.service.ts (hunk 14 above)
Comment (Simperfit) [inline src/app/hm-admin/services/schema/schema.service.ts:81]: But we will do that in another PR, this is not in the scope of this one.
